plugins_by_families now matches with algorithms

p4vv37 opened the pull request:
Updated logic.plugins_by_families, now it uses the matching algorithm of
the plugin,
Updated test of plugins_by_families

p4vv37 committed:
plugins_by_families now matches with algorithms

Updated logic.plugins_by_families, now it uses the matching algorithm of
the plugin,
Updated test of plugins_by_families

mottosso commented:
Thanks @p4vv37, would it be possible to provide an example of what didn't work before? A little runnable snippet of code.

p4vv37 commented:
Sure, from changed tests:

    class ClassD(pyblish.api.Collector):
        families = ["a", "b"]
        match = api.Intersection

    class ClassE(pyblish.api.Collector):
        families = ["a", "b"]
        match = api.Subset

    class ClassF(pyblish.api.Collector):
        families = ["a", "b"]
        match = api.Exact

    assert pyblish.logic.plugins_by_families(
        [ClassD, ClassE, ClassF], ["a"]) == [ClassD]

    assert pyblish.logic.plugins_by_families(
        [ClassD, ClassE, ClassF], ["a", "b"]) == [ClassD, ClassE, ClassF]

    assert pyblish.logic.plugins_by_families(
        [ClassD, ClassE, ClassF], ["a", "b", "c"]) == [ClassD, ClassE]

p4vv37 commented:
Sorry, I've just missclicked.

The snipplet of code that runs after change and did not run before it:


    class ClassD(pyblish.api.Collector):
        families = ["a", "b"]
        match = api.Intersection
 
    class ClassE(pyblish.api.Collector):
        families = ["a", "b"]
        match = api.Subset

    class ClassF(pyblish.api.Collector):
        families = ["a", "b"]
        match = api.Exact

    assert pyblish.logic.plugins_by_families(
        [ClassD, ClassE, ClassF], ["a"]) == [ClassD]

    assert pyblish.logic.plugins_by_families(
        [ClassD, ClassE, ClassF], ["a", "b"]) == [ClassD, ClassE, ClassF]

    assert pyblish.logic.plugins_by_families(
        [ClassD, ClassE, ClassF], ["a", "b", "c"]) == [ClassD, ClassE]


I've included it to the pyblish-base\tests\pre13\__init__.py

mottosso commented:
Cool, I'll have a closer look at this as soon as I can.

About the `preXX` folders and files, they are for tests of older versions of Pyblish; in this case, `pre13` means it includes tests for version 1.3 of Pyblish. They aren't anything that should ever change, as that would mean that newer functionality breaks backwards compatibility.

Would it be possible to put these tests outside of the `pre13` directory?

mottosso commented:
Just so I'm on the same page, what triggered the need for this functionality from you? What did you try to do, where you noticed this was a problem?

If you could post a usage example of what you couldn't do before, but could do with this feature, that would be most helpful!

p4vv37 committed:
moved tests, reverted tst_logic from pre13

p4vv37 commented:
I've reverted the test from pre13 and moved this test.

I'm using this funcionality in my modifications to pyblish-lite for hiding plugins incompatible with checked instances:
https://github.com/p4vv37/pyblish-lite/tree/my_branch

I thought, that it would be wiser to use this function, instead of writing my own, to check if a plugin should be hidden.

mottosso commented:
> I thought, that it would be wiser to use this function, instead of writing my own, to check if a plugin should be hidden.

The thinking sounds right, thanks @p4vv37! I'll take a look at this as soon as I can.

mottosso commented on pyblish/logic.py:
```diff
@@ -153,10 +153,26 @@ def plugins_by_families(plugins, families):
 
     """
 
+    algorithms = {
```
Since the same algorithms are being used in both this function, and `instances_by_plugin`, what do you think about moving this dictionary to the module-level, and prefix it with an underscore to highlight the fact that it is internal. That way, we could use it in both functions, and changing it would be simpler and more predictable.

p4vv37 replied:
Right, that's a nice idea. I'm committing the change.

mottosso reviewed (changes requested):
Great work on this, the tests look solid, style is consistent and functionality superb.

Have a look at the one note I made, and let me know what you think. This is nearly ready for merge and release.

Thanks again @p4vv37!

p4vv37 committed:
Changed algorithms dict to module-level _algorithms

As suggested by @mottosso _algorithms is now a module-level dictionary
and both functions uses it

mottosso commented:
This looks good to me. Happy for me to merge this?

p4vv37 commented:
Sure, go ahead
